### PR TITLE
make Libtask compatible with Julia 1.6

### PR DESCRIPTION
**Pull request by KDr2**
Not ready to merge until https://github.com/JuliaPackaging/Yggdrasil/pull/2176 is merged.

### CODE REVIEW

File: src/ctask.jl [hunk 1]
```diff
@@ -88,7 +94,7 @@ end
 
 function Base.copy(ctask::CTask)
     task = ctask.task
-    state = getproperty(task, :state)
+    state = task.state
     if state !== :runnable && state !== :done
```
Comment (devmotion) [inline src/ctask.jl:98]: Maybe just
```suggestion
    if istaskfailed(task)
```
?

File: src/ctask.jl [hunk 2]
```diff
@@ -139,7 +145,7 @@ function produce(v)
     end
 
     # Internal check to make sure that it is possible to switch to the consumer.
-    @assert getproperty(task, :state) in (:runnable, :queued)
+    @assert task.state in (:runnable, :queued)
 
     @static if VERSION < v"1.1.9999"
```
Comment (devmotion) [inline src/ctask.jl:150]: Remove this branch?
```suggestion
```

File: src/ctask.jl [hunk 3]
```diff
@@ -139,7 +145,7 @@ function produce(v)
     end
 
     # Internal check to make sure that it is possible to switch to the consumer.
-    @assert getproperty(task, :state) in (:runnable, :queued)
+    @assert task.state in (:runnable, :queued)
```
Comment (devmotion) [inline src/ctask.jl:148]:
```suggestion
    @assert !istaskdone(state) && !istaskfailed(task)
```

File: src/ctask.jl [hunk 4]
```diff
@@ -37,7 +42,7 @@ function n_copies(t::Task)
 end
 
 function enable_stack_copying(t::Task)
-    state = getproperty(t, :state)
+    state = t.state
     if state !== :runnable && state !== :done
```
Comment (devmotion) [inline src/ctask.jl:46]: Maybe just
```suggestion
    if istaskfailed(t)
```

File: src/ctask.jl (hunk 1 above)
Comment (devmotion) [reply src/ctask.jl:98]: BTW `_istaskfailed` can be removed since we only support Julia >= 1.3.

File: src/ctask.jl [hunk 5]
```diff
@@ -203,7 +209,7 @@ function consume(ctask::CTask, values...)
         push!(producer.storage[:consumers].waitq, ct)
     end
 
-    if getproperty(producer, :state) === :runnable
+    if producer.state === :runnable
```
Comment (devmotion) [inline src/ctask.jl:212]: Maybe also here use something that is less dependent on the internal state?
```suggestion
    if !istaskdone(producer) && !istaskfailed(producer)
```

File: src/ctask.jl (hunk 2 above)
Comment (KDr2) [reply src/ctask.jl:150]: Done.

File: src/ctask.jl (hunk 5 above)
Comment (KDr2) [reply src/ctask.jl:212]: There's another potential state (`:queued`) here, so I think using `=== :runnable` is more precise here.

File: src/ctask.jl (hunk 1 above)
Comment (KDr2) [reply src/ctask.jl:98]: Done.

File: src/ctask.jl (hunk 5 above)
Comment (devmotion) [reply src/ctask.jl:212]: I think it was removed already before Julia 1.3, so it should not matter anymore.

Comment (devmotion) [reply src/ctask.jl:212]: I couldn't find it in https://github.com/JuliaLang/julia/blob/v1.3.0-alpha/base/task.jl.

Comment (devmotion) [reply src/ctask.jl:212]: Another indication is this check in ctask.jl: https://github.com/TuringLang/Libtask.jl/blob/a664d3dc5ee8e56ab85dc9368e94f68c8b150083/src/ctask.jl#L144-L146 Apparently it was a thing only in Julia < 1.2.

File: src/ctask.jl (hunk 4 above)
Comment (KDr2) [reply src/ctask.jl]: Done.

File: src/ctask.jl (hunk 5 above)
Comment (KDr2) [reply src/ctask.jl]: You are right, updated.

**Review by devmotion — COMMENTED** [review]
Looks good to me but I guess we have to wait until Julia 1.6 is released (or at least one can build binaries for Julia 1.6 on Yggdrasil).

On the other hand, it does not break anything on Julia < 1.6 and only improves the code, so we could also merge it now. It is just not guaranteed that these fixes are sufficient for Julia 1.6.

**Comment by KDr2** [comment]
> On the other hand, it does not break anything on Julia < 1.6 and only improves the code, so we could also merge it now.

Agree.

>  It is just not guaranteed that these fixes are sufficient for Julia 1.6.

I ran the unit tests locally (on linux) with the updated C source, all of them passed.

So it seems that this PR can be merged now.

**Comment by devmotion** [comment]
Good I trust you :slightly_smiling_face: And if we notice any problems in the future, we can always make a bugfix release.

I guess it would be useful to make a release right away if people want to use it with Julia master (even though I don't think it's very likely since they would have to build everything manually)?

**Review by devmotion — APPROVED** [review]

**Review by yebai — APPROVED** [review]
Great work, thanks @KDr2!

**Pull request merged**